### PR TITLE
Add Manga Toshokan Z

### LINKED ISSUE #2343: マンガ図書館Z / Manga Toshokan Z
### Source name

マンガ図書館Z / Manga Toshokan Z

### Source link

https://www.mangaz.com/

### Source language

Japanese

### Other details

Has a lot of free manga to read

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by sinkableShip**
Closes #2343

there's r18 manga section but it all trial version should we even add it? i already add it just in case but it might be uncomfortable to read cut short manga

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/Crypto.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.security.KeyPairGenerator
+import java.security.PrivateKey
+import java.security.PublicKey
+import java.security.spec.RSAKeyGenParameterSpec
+import javax.crypto.Cipher
+
+private val json: Json by injectLazy()
+
+private const val BEGIN_PUBLIC_KEY = "-----BEGIN PUBLIC KEY-----"
+private const val END_PUBLIC_KEY = "-----END PUBLIC KEY-----"
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/Crypto.kt]: inline these

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    private fun r18Interceptor(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val response = chain.proceed(request)
+
+        // open access to R18 section
+        if (request.url.host == "r18.mangaz.com" && isR18.not()) {
+            val url = baseUrl.toHttpUrl().newBuilder()
+                .host("r18.mangaz.com")
+                .addPathSegments("attention/r18/yes")
+                .build()
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]:
```suggestion
            val url = "https://r18.mangaz.com/attention/r18/yes"
```

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+            val r18Request = Request.Builder()
+                .url(url)
+                .head()
+                .build()
+
+            isR18 = true
+            client.newCall(r18Request).execute()
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]:
```suggestion
            client.newCall(r18Request).execute().close()
```

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        getSerial()
+    }
+
+    private var isR18 = false
+
+    private fun r18Interceptor(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val response = chain.proceed(request)
+
+        // open access to R18 section
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt:49]: shouldn't it be done before getting the response?

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+            isR18 = true
+            client.newCall(r18Request).execute()
+        }
+
+        return response
+    }
+
+    override fun popularMangaRequest(page: Int): Request {
+        val url = baseUrl.toHttpUrl().newBuilder()
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]: generally when you don't have anything to encode in the url (like user provided query), hardcoding it as string is fine

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        }
+
+        return GET(url.build(), headers)
+    }
+
+    override fun searchMangaParse(response: Response): MangasPage {
+        val document = response.asJsoup()
+        val mangas = document.selectFirst(".itemList")!!.children().filter { child ->
+            child.`is`("li")
+        }.mangasFromListElements()
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]: why not ".itemList li" or ".itemList > li" ?

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+                thumbnail_url = if (img.hasAttr("src")) {
+                    img.attr("src")
+                } else {
+                    img.attr("data-src")
+                }
+
+                status = when {
+                    li.selectFirst(".iconContinues") != null -> SManga.ONGOING
+                    li.selectFirst("iconEnd") != null -> SManga.COMPLETED
+                    else -> { SManga.UNKNOWN }
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]:
```suggestion
                    else -> SManga.UNKNOWN
```

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+
+        val ticketRequest = Request.Builder()
+            .url(ticketUrl)
+            .headers(headers)
+            .head()
+            .build()
+
+        runCatching {
+            client.newCall(ticketRequest).execute()
+        }.getOrNull() ?: throw Exception("Fail to retrieve ticket")
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]:
```suggestion
        try {
            client.newCall(ticketRequest).execute().close()
        } catch (_: Exception) {
            throw Exception("Fail to retrieve ticket")
        }
```

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+    }
+
+    private fun getSerial(): String {
+        val url = virgoBuilder()
+            .addPathSegment("app.js")
+            .build()
+
+        val response = runCatching {
+            client.newCall(GET(url, headers)).execute()
+        }.getOrNull() ?: throw Exception("Fail to retrieve serial")
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]:
```suggestion
        val response = try {
            client.newCall(GET(url, headers)).execute()
        } catch (_: Exception) {
            throw Exception("Fail to retrieve serial")
        }
```

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+            Page(i, imageUrl = imageUrl.toString())
+        }
+    }
+
+    override fun imageUrlParse(response: Response): String {
+        throw UnsupportedOperationException()
+    }
+
+    companion object {
+        const val LATEST_MANGA_COUNT_PER_PAGE = 50
```
Comment (AwkwardPeak7) [inline src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]: inline this

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt (hunk 6 above)
Comment (sinkableShip) [reply src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt]: there can be multiple .itemList, like in here:
https://www.mangaz.com/mens/title/index?category=mens&query=%E3%82%8F%E3%81%9F%E3%81%97&sort=new&search=input

File: src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt (hunk 4 above)
Comment (sinkableShip) [reply src/ja/mangatoshokanz/src/eu/kanade/tachiyomi/extension/ja/mangatoshokanz/MangaToshokanZ.kt:49]: ahh right, looks like i forgot to change it

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**